packages/server/src/fhir/operations/resourcegraph.ts

```diff
@@ -0,0 +1,333 @@
+  PropertyType,
+  toTypedValue,
+  TypedValue,
+} from '@medplum/core';
+import {
+  Bundle,
+  GraphDefinition,
+  GraphDefinitionLink,
+  GraphDefinitionLinkTarget,
+  Reference,
+  Resource,
+} from '@medplum/fhirtypes';
+import { Request, Response } from 'express';
+
+import { Repository } from '../repo';
+import { sendResponse } from '../routes';
+
+/**
+ * Handles a Resource $graph request.
+ *
+ * The operation fetches all the data related to this resources as defined by a GraphDefinition resource
+ *
+ * See: https://hl7.org/fhir/plandefinition-operation-apply.html
+ *
+ * @param req The HTTP request.
+ * @param res The HTTP response.
+ */
+export async function resourceGraphHandler(req: Request, res: Response): Promise<void> {
+  const { resourceType, id } = req.params;
+  const repo = res.locals.repo as Repository;
+
+  const rootResource = await repo.readResource(resourceType, id);
+
+  const definition = await validateQueryParameters(req, res);
+
+  if (!definition) {
+    return;
+  }
+
+  if (!definition.start || definition.start !== resourceType) {
+    throw new OperationOutcomeError(badRequest('Missing or incorrect `start` type'));
+  }
+  const results = [rootResource] as Resource[];
+  const resourceCache = {} as Record<string, Resource>;
+  resourceCache[getReferenceString(rootResource)] = rootResource;
+  if ((rootResource as any).url) {
+    resourceCache[(rootResource as any).url] = rootResource;
+  }
+  await followLinks(rootResource, definition.link, results, resourceCache, repo);
+
+  sendResponse(res, allOk, {
+    resourceType: 'Bundle',
+    entry: deduplicateResources(results).map((r) => ({
+      resource: r,
+    })),
+    type: 'collection',
+  } as Bundle);
+}
+
+/**
+ *
+ * current resource is the start resource
+ * for each link:
+ *  evaluate the path
+ *  assert we get a collection of references
+ *  filter the references to have the target resourceType
+ *  for each elem in the collection:
+ *    read the reference
+ *    add the result to the bundle
+ *    follow the subLink on the element
+ * @param resource
+ * @param path
+ */
+type FhirPathLink = GraphDefinitionLink & Required<Pick<GraphDefinitionLink, 'path'>>;
+function isFhirPathLink(link: GraphDefinitionLink): link is FhirPathLink {
+  return link.path !== undefined;
+}
+
+type SearchLink = Omit<GraphDefinitionLink, 'path'> & {
+  target: Required<Pick<GraphDefinitionLinkTarget, 'type' | 'params'>>;
+};
+function isSearchLink(link: GraphDefinitionLink): link is SearchLink {
+  return link.path === undefined && link.target !== undefined && link.target.every((t) => t.type && t.params);
+}
+async function followLinks(
+  resource: Resource,
+  links: GraphDefinitionLink[] | undefined,
+  results: Resource[],
+  resourceCache: Record<string, Resource>,
+  repo: Repository,
+  depth: number = 0
+) {
+  // Circuit Breaker
+  if (results.length > 1000 || depth >= 5) {
+    return;
+  }
+
+  // Recursive Base Case
+  if (!links || links.length == 0) {
```

Comment:
You can omit the `links.length == 0` check, because it is handled by an empty `for` loop.

Also, we try to always use `===`, never `==`.